### PR TITLE
add an optional builder pattern as well as extension traits for tags when using the builder interface

### PR DESCRIPTION
**Pull request by ahl**
@karencfv can you take a look at the nexus builder output? https://github.com/oxidecomputer/progenitor/pull/86/files#diff-e118427cd84c513809413bad47cd7e2f4edb8eae1dfad97418a8c1319f437d17

@jclulow tagging you in case you're interested

### CODE REVIEW

**Comment by karencfv** [comment]
Oh nice!!!!! I'll take a look today :)

**Review by karencfv — COMMENTED** [review]
Thanks for this @ahl !! It looks really great and easy to work with.

I see that there'll still be some work to be done in regards to standardising the field descriptions (or lack thereof 😄 ), and struct naming. Although, that's not relevant to this PR in specific.

File: progenitor-impl/tests/output/nexus-builder-tagged.out [hunk 1]
```diff
@@ -0,0 +1,11106 @@
...
+    fn project_images_get_image(&self) -> builder::ProjectImagesGetImage {
+        builder::ProjectImagesGetImage::new(self)
+    }
+
+    fn project_images_delete_image(&self) -> builder::ProjectImagesDeleteImage {
+        builder::ProjectImagesDeleteImage::new(self)
+    }
+}
+
+pub trait ClientImagesGlobalExt {
```
[the remark this answers is not in the document]
Comment (jclulow) [reply progenitor-impl/tests/output/nexus-builder-tagged.out:1797]: I'm not sure if this pattern is my fault or if I lifted it from somewhere else, but when I have used it, it's been meant to be a trait that **ext**ends something with **ext**ra methods.  So, either Extended or Extra.

**Comment by jclulow** [comment]
I brought [my `jclulow` branch](https://github.com/oxidecomputer/progenitor/commits/jclulow) up to date with `main`, which is what I currently use for `buildomat`, and then rebased this PR on top to get the builder stuff.

The first thing I noticed is that `reqwest::Body` is not `Clone`, so the `#[derive(Clone)]` on some of the types doesn't work:

```
jclulow@vulcan /ws/buildomat
 $ cargo check
    Checking buildomat-openapi v0.0.0 (/ws/buildomat/openapi)
error[E0277]: the trait bound `Body: Clone` is not satisfied
    --> openapi/src/lib.rs:1916:9
     |
1912 |     #[derive(Clone)]
     |              ----- in this derive macro expansion
...
1916 |         body: Option<reqwest::Body>,
     |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Clone` is not implemented for `Body`
     |
     = note: required because of the requirements on the impl of `Clone` for `std::option::Option<Body>`
     = note: this error originates in the derive macro `Clone` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `Body: Clone` is not satisfied
    --> openapi/src/lib.rs:2797:9
     |
2793 |     #[derive(Clone)]
     |              ----- in this derive macro expansion
...
2797 |         body: Option<reqwest::Body>,
     |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Clone` is not implemented for `Body`
     |
     = note: required because of the requirements on the impl of `Clone` for `std::option::Option<Body>`
     = note: this error originates in the derive macro `Clone` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0277`.
error: could not compile `buildomat-openapi` due to 2 previous errors
```

i.e., for types like:

```rust
    #[derive(Clone)]
    pub struct JobUploadChunk<'a> {
        client: &'a super::Client,
        job: Option<String>,
        body: Option<reqwest::Body>,
    }
```

I'm tinkering with the generated output though, and I'll report back on how that goes.

File: progenitor-impl/tests/output/nexus-builder-tagged.out (hunk 1 above)
[the remark this answers is not in the document]
Comment (karencfv) [reply progenitor-impl/tests/output/nexus-builder-tagged.out:1797]: ah! gotcha thanks

**Comment by jclulow** [comment]
A few more thoughts:
* It would be good if builder functions took, say, `AsRef<str>`, instead of the owned `String`:
  ```rust
  let blah = "blah".to_string();
  /*
   * Today you have to pass an owned structure, so you would have to:
   */
  client.endpoint().blah(blah.to_string()).send().await?;
  client.endpoint().blah(blah).send().await?;
  /*
   * But with AsRef<str> I think you could then, say:
   */
  client.endpoint().blah(&blah).send().await?;
  client.endpoint().blah(&blah).send().await?;
  client.endpoint().blah(blah).send().await?;
  ```

* Initial experience with the builder pattern stuff is neat!  It seems to mostly cover query arguments, though, and I wonder what we should do about body types.  So far, with the PR as-is, I have to:
  ```rust
  let res = c
      .client
      .factory_worker_append()
      .worker(i.worker)
      .body(FactoryWorkerAppend {
          payload: ie.payload.to_string(),
          stream: ie.stream.to_string(),
          time: ie.time.0,
      })
      .send()
      .await?;
  ```

  I reckon it'd be good to be able to, instead, do one of these (or some third thing?):
  * Combine the properties from the body type into the top-level builder (or, I guess, fail if they conflict in name with query parameters?):
    ```rust
    let res = c
        .client
        .factory_worker_append()
        .worker(&i.worker)
        .payload(&ie.payload)
        .stream(&ie.stream)
        .time(ie.time.0)
        .send()
        .await?;
    ```
  * Create separate builders for the body type, which is I think mostly what the AWS SDK does today:
    ```rust
    let res = c
        .client
        .factory_worker_append()
        .worker(i.worker)
        .body(
            FactoryWorkerAppend::new()
                .payload(&ie.payload)
                .stream(&ie.stream)
                .time(ie.time.0)
                .build()
        )
        .send()
        .await?;
    ```

File: progenitor-impl/tests/output/nexus-builder-tagged.out (hunk 1 above)
[the remark this answers is not in the document]
Comment (ahl) [reply progenitor-impl/tests/output/nexus-builder-tagged.out:1797]: This is semi-standard in rust and I think they're called "extension" traits e.g.

https://docs.rs/futures/latest/futures/stream/index.html#traits

[the remark this answers is not in the document]
Comment (karencfv) [reply progenitor-impl/tests/output/nexus-builder-tagged.out:1797]: 🙇‍♀️

**Comment by ahl** [comment]
Thanks for taking a look @jclulow 

> * It would be good if builder functions took, say, `AsRef<str>`, instead of the owned `String`[..]

I'm all for it. I'm not sure what to do with string-like wrapper such as the `Name` type from nexus. Do you have any suggestions?

>   * Combine the properties from the body type into the top-level builder (or, I guess, fail if they conflict in name with query parameters?)

I like this idea, but I'm torn. Is it right to think of the properties of the body parameter as arguments to the operation? I kind of think of the body itself as the parameter. Conversely, I don't love that there's a parameter called `body` in that I think it's nice to abstract away the HTTP-ness as much as possible. Have you seen examples to inform one direction or the other?

>   * Create separate builders for the body type, which is I think mostly what the AWS SDK does today:

I like this very much... not least of all because it seems deferrable from this PR and a strict improvement on the types we're already generating. Perhaps `builderify` should be a configuration option to `typify`.

**Comment by karencfv** [comment]
Woohoo merged! Going to try this out with the Rust SDK/CLI and get that that refactoring (rewrite? 😅 ) work started.

**Comment by ahl** [comment]
@karencfv just a heads up that there are a couple of known issues preventing generation from nexus.json

**Comment by karencfv** [comment]
Ah! gotcha, are those two on here https://github.com/oxidecomputer/progenitor/issues so I can follow them?

**Pull request merged**